docs/src/userguide/extension_types.rst

```diff
@@ -484,6 +484,76 @@ statically sized freelist of ``N`` instances for a given type.  Example::
     penguin = Penguin('fish 2')  # does not need to allocate memory!
 
 
+.. _existing-pointers-instantiation:
+
+Instantiation from existing C/C++ pointers
+===========================================
+
+It is quite common to want to instantiate an extension class from an existing
+(pointer to a) data structure, often as returned by external C/C++ functions.
+
+As extension classes can only accept Python objects as arguments in their
+contructors, this necessitates the use of factory functions. For example, ::
+
+    from libc.stdlib cimport malloc, free
+
+    # Example C struct
+    ctypedef struct my_c_struct:
+        int a
+        int b
+
+
+    cdef class WrapperClass:
+        """A wrapper class for a C/C++ data structure"""
+        cdef my_c_struct *_ptr
+
+        def __cinit__(self):
+            # On cinit, do not create new structure but set pointer to NULL
+            self._ptr = NULL
+
+        def __dealloc__(self):
+            # De-allocate if we have a non-null pointer
+            if self._ptr is not NULL:
+                free(self._ptr)
+
+
+    cdef WrapperClass PyWrapperClass(my_c_struct *_ptr):
```

Comment:
These could also be static members on the class itself, which might be more natural than free functions.

Reply:
I don't think static methods would be appropriate in this case. `PyWrapperClass` could well be a `cpdef` function to be used in both C and Python space, which static methods do not currently allow. Making it a regular python function then results in it having to be called via Python even when used in C code.

Happy to add a note that they could be written as `@staticmethods` if you'd like, depending on developer preference and how they will be used.

Reply:
It can't be a cpdef method, for the same reason it can't just be the constructor, it must be a cdef method. I was suggesting something like

```
cdef class WrapperClass:
    ...
    @staticmethod
    cdef WrapperClass create(c_data *ptr):
        ...


cdef WrapperClass o = WrapperClass.create(&c_data)
```

Reply:
Sure, will make that change.